### ISSUE TITLE
Partial update failing

### ISSUE THREAD

**Issue by darkmoon03** [logstash-plugins/logstash-output-elasticsearch#789]
https://discuss.elastic.co/t/version-conflict-document-already-exists-current-version-1/139230

v5.6.10 (and v5.6.9), updated all logstash plugins
CenOS 7

With this config:
```
output {
   if ([type] == "state" ) {
      elasticsearch {
         hosts => [<hosts> ]
         index => "%{[meta][target][index]}"
         document_id => "%{[@metadata][target][id]}"
         action => "update"
         doc_as_upsert => true
         template_overwrite => false
         manage_template => false
         retry_on_conflict => 5
     }
  }
}
```

I get this error on any update (creates work):
[2018-07-09T15:10:44.971-0400][WARN ][logstash.outputs.elasticsearch] Failed action. {:status=>409, :action=>["update", {:_id=>"f4:4d:30:60:8a:31", :_index=>"state_mac", :_type=>"state", :_routing=>nil, :_retry_on_conflict=>1}, 2018-07-09T19:09:45.000Z %{host} %{message}], :response=>{"update"=>{"_index"=>"state_mac", "_type"=>"state", "_id"=>"f4:4d:30:60:8a:31", "status"=>409, "error"=>{"type"=>"version_conflict_engine_exception", "reason"=>"[state][f4:4d:30:60:8a:31]: version conflict, document already exists (current version [1])", "index_uuid"=>"huFaDcR5RgeG92F5S8F9kw", "shard"=>"2", "index"=>"state_mac"}}}}

I know the document already exists, it's an update, not a create. I've played around with retries and various version settings without success.

Anyone have any ideas on how to disable the version check? It shouldn't even be checking. The docs (https://www.elastic.co/blog/elasticsearch-versioning-support) say it's optional (and off by default), but not how to disable it.

Everything works otherwise. Updates using the elastic update api (via curl) work. This config works in v5.4.   This looks like an output plugin bug to me.

**Comment by darkmoon03** [comment]
Does anyone have a working 5.6 config that does partial updates (update/upsert)?

This works in 5.4 perfectly.   It still works via the API (curl).  Maybe one of the options has changed?

The 5.x and 6.x documentation both say that version checking is optional, and not active unless turned on.  This looks like a bug in the logstash elasticsearch output plugin.

**Comment by darkmoon03** [comment]
I know this is a rare use case, but can someone please take a look at this?

This is a documented feature and it's not working.

**Comment by darkmoon03** [comment]
I know this is a rare use case, but can someone please take a look at this?

This is a documented feature and it's not working.

**Comment by mlasak** [comment]
No solution to this? I'm having the same trouble.

**Comment by darkmoon03** [comment]
Sorry, no.  Yours is the first response I've gotten here or in the forums.  (I've gotten some feedback in IRC from other users, but no one there was attempting this particular trick.)

**Comment by mlasak** [comment]
I believe it is simply broken. I gave up doing "update", instead using "index" as action and do some post-processing after every event landed safely in the elasticsearch cluster. Thanks for your response.

**Comment by darkmoon03** [comment]
I dropped back to 'index' as well, but that doesn't work in the cases where I need to do a partial update, rather than a full replacement.

**Comment by darkmoon03** [comment]
Documentation says you can do it, and it works from the API (curl), so yes, it's broken.

**Comment by darkmoon03** [comment]
Bump to keep this from expiring.

**Comment by darkmoon03** [comment]
Bump to keep this from expiring.

**Comment by darkmoon03** [comment]
Weekly bump.  Please, will someone take a look at this bug?

**Comment by jsvd** [comment]
This is strange, on the log entry it shows `_retry_on_conflict` to be 1 while the logstash config shows `retry_on_conflict => 5`.

For updates elasticsearch uses internal versioning to decide if there's a conflict, and that only kicks in if the version parameter is set in the request. This plugin [only sets the version if the `version` parameter is set](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/v7.4.0/lib/logstash/outputs/elasticsearch/common.rb#L189).

Can we get a tcpdump to see the actual bulk request that logstash is executing against elasticsearch?

**Comment by darkmoon03** [comment]
Yup, it's a bug.  Version is not set, but it's doing the check anyway.

I'm pretty sure a tcpdump won't help - we're using certs and ssl.  Our config is posted above.  Removing the retry_on_conflict doesn't help.

Thanks for the reply by the way, I've been trying to get some traction on this for a good too months.

**Comment by jsvd** [comment]
This is my test:

```
input {
  stdin { codec => json }
}
output {
  elasticsearch {
     index => "test"
     document_id => "%{[id]}"
     action => "update"
     doc_as_upsert => true
     manage_template => false
     retry_on_conflict => 5
  }
}
```

I started logstash and then sent 3 documents with the same id and after each document I checked elasticsearch for the document:

```
/tmp/logstash-6.4.0 % bin/logstash -f cfg
[2018-08-29T15:38:21,742][INFO ][logstash.runner          ] Starting Logstash {"logstash.version"=>"6.4.0"}
[2018-08-29T15:38:24,354][INFO ][logstash.pipeline        ] Starting pipeline {:pipeline_id=>"main", "pipeline.workers"=>4, "pipeline.batch.size"=>125, "pipeline.batch.delay"=>50}
[2018-08-29T15:38:25,017][INFO ][logstash.outputs.elasticsearch] Elasticsearch pool URLs updated {:changes=>{:removed=>[], :added=>[http://127.0.0.1:9200/]}}
[2018-08-29T15:38:25,033][INFO ][logstash.outputs.elasticsearch] Running health check to see if an Elasticsearch connection is working {:healthcheck_url=>http://127.0.0.1:9200/, :path=>"/"}
[2018-08-29T15:38:25,370][WARN ][logstash.outputs.elasticsearch] Restored connection to ES instance {:url=>"http://127.0.0.1:9200/"}
[2018-08-29T15:38:25,512][INFO ][logstash.outputs.elasticsearch] ES Output version determined {:es_version=>6}
[2018-08-29T15:38:25,588][INFO ][logstash.outputs.elasticsearch] New Elasticsearch output {:class=>"LogStash::Outputs::ElasticSearch", :hosts=>["//127.0.0.1"]}
[2018-08-29T15:38:25,716][INFO ][logstash.inputs.stdin    ] Automatically switching from json to json_lines codec {:plugin=>"stdin"}
The stdin plugin is now waiting for input:
[2018-08-29T15:38:25,776][INFO ][logstash.pipeline        ] Pipeline started successfully {:pipeline_id=>"main", :thread=>"#<Thread:0x72c0fded run>"}
[2018-08-29T15:38:25,845][INFO ][logstash.agent           ] Pipelines running {:count=>1, :running_pipelines=>[:main], :non_running_pipelines=>[]}
[2018-08-29T15:38:26,244][INFO ][logstash.agent           ] Successfully started Logstash API endpoint {:port=>9600}
{"id": 33, "message": "hello"}
{"id": 33, "message": "ola"}  
{"id": 33, "message": "おはよう"}
```

```
% curl -s http://localhost:9200/test/doc/33 | jq
{
  "_index": "test",
  "_type": "doc",
  "_id": "33",
  "_version": 1,
  "found": true,
  "_source": {
    "message": "hello",
    "host": "Joaos-MacBook-Pro-5.local",
    "@timestamp": "2018-08-29T14:38:43.361Z",
    "id": 33,
    "@version": "1"
  }
}
% curl -s http://localhost:9200/test/doc/33 | jq
{
  "_index": "test",
  "_type": "doc",
  "_id": "33",
  "_version": 2,
  "found": true,
  "_source": {
    "message": "ola",
    "host": "Joaos-MacBook-Pro-5.local",
    "@timestamp": "2018-08-29T14:39:17.567Z",
    "id": 33,
    "@version": "1"
  }
}
% curl -s http://localhost:9200/test/doc/33 | jq
{
  "_index": "test",
  "_type": "doc",
  "_id": "33",
  "_version": 3,
  "found": true,
  "_source": {
    "message": "おはよう",
    "host": "Joaos-MacBook-Pro-5.local",
    "@timestamp": "2018-08-29T14:40:15.724Z",
    "id": 33,
    "@version": "1"
  }
}
```

**Comment by jsvd** [comment]
I tried this with both 6.4.0 and 5.6.11

**Comment by darkmoon03** [comment]
What happens if you try partial updates?  (have a  field 'foo' in your first record, 'bar' in your second, and 'baz' in your third.)

I can get overwrites to work, but not partial updates.

**Comment by jsvd** [comment]
I tried:

```
{"id": 14, "message": "hello"}
{"id": 14, "doc": { "test": "1"} }
{"id": 14, "new_field": "something"}
```

and got: 

```
% curl -s http://localhost:9200/test/doc/14 | jq
{
  "_index": "test",
  "_type": "doc",
  "_id": "14",
  "_version": 1,
  "found": true,
  "_source": {
    "host": "Joaos-MacBook-Pro-5.local",
    "message": "hello",
    "@timestamp": "2018-08-29T15:52:23.993Z",
    "id": 14,
    "@version": "1"
  }
}
% curl -s http://localhost:9200/test/doc/14 | jq
{
  "_index": "test",
  "_type": "doc",
  "_id": "14",
  "_version": 2,
  "found": true,
  "_source": {
    "host": "Joaos-MacBook-Pro-5.local",
    "message": "hello",
    "@timestamp": "2018-08-29T15:52:58.258Z",
    "id": 14,
    "@version": "1",
    "doc": {
      "test": "1"
    }
  }
}
% curl -s http://localhost:9200/test/doc/14 | jq
{
  "_index": "test",
  "_type": "doc",
  "_id": "14",
  "_version": 3,
  "found": true,
  "_source": {
    "host": "Joaos-MacBook-Pro-5.local",
    "message": "hello",
    "@timestamp": "2018-08-29T15:53:42.221Z",
    "id": 14,
    "@version": "1",
    "doc": {
      "test": "1"
    },
    "new_field": "something"
  }
}
```

**Comment by darkmoon03** [comment]
Thank you.

I'll try that exact example (I'll need to adjust my test environment).

**Comment by darkmoon03** [comment]
Still not working.  If I use a file input and append entries to the file (your example), it seems to work, but not for real data.

[2018-08-29T14:34:47.807-0400][WARN ][logstash.outputs.elasticsearch] Failed action. {:status=>409, :action=>["update", {:_id=>"00:31:46:17:df:f0", :_index=>"log_test", :_type=>"state", :_routing=>nil, :_retry_on_conflict=>5}, 2018-08-29T18:34:36.000Z [] %{message}], :response=>{"update"=>{"_index"=>"log_test", "_type"=>"state", "_id"=>"00:31:46:17:df:f0", "status"=>409, "error"=>{"type"=>"version_conflict_engine_exception", "reason"=>"[state][00:31:46:17:df:f0]: version conflict, document already exists (current version [1])", "index_uuid"=>"uqU84VkaQpmq-g6yFTAMVA", "shard"=>"0", "index"=>"log_test"}}}}

Here's the current config:
elasticsearch {
            hosts => [<>]
            index => "log_test"
            codec => json
            document_id => "%{[@metadata][target][id]}"
            action => "update"
            doc_as_upsert => true
            id => "logfilter-dev-01.internal.cls.vt.edu_es_state"
            ssl => true
            ssl_certificate_verification => false
            keystore => <>
            keystore_password => <>
            cacert => <>
            template_overwrite => false
            manage_template => false
            retry_on_conflict => 5
        }

**Comment by darkmoon03** [comment]
( /sigh @version is all lower case.  The autoformatting is causing issues again.)

**Comment by jsvd** [comment]
`@version` is just a field in the event, that gets sent in the payload part of the bulk request that represents the content of the document. The version field that Elasticsearch uses for internal versioning checks is set in the [action and metadata line](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-bulk.html) of the bulk request. 
This second version parameter is only set if you pass `version => N` in the es output on logstash. So this should not be raising a 409..
(You can see in my tests that es returns documents where  `@version: 1` is present in the _source field)

Which version of Elasticsearch is this?

**Comment by mlasak** [comment]
I am encountering the issue with version 6.3.2 of both Elasticsearch and Logstash.

**Comment by jsvd** [comment]
@mlasak can you post your es output config and the logstash startup logs with --log.level=debug?

**Comment by darkmoon03** [comment]
Elastic and logstash versions 5.6.10

**Comment by jsvd** [comment]
I still can't seem to replicate this. is anyone not using ssl (or able to reproduce in dev without ssl) that can use tcpdump/wireshark to show the actual http request/response happening?

**Comment by mlasak** [comment]
I do not use ssl. Will provide you the logs asap (travelling now)

**Comment by jsvd** [comment]
>  is this a magic number that will work for all scenarios?

Not really. Like you said, it for a high frequency concurrent update rate on a single document it may make sense to increase the value.

There is a good [discuss thread conversation](https://discuss.elastic.co/t/whats-appropriate-value-at-retry-on-conflict/54312/3) where one of the ES devs chimes in about choosing the value for `retry_on_conflict`.

**Comment by darkmoon03** [comment]
Setting retry_on_conflict doesn't work for my production data - it just fails that number of times, even when I send in a single document (no concurrency issues).

I did get the simple example you posted above to work, but I don't see what the difference is.  I was hoping you'd see something in the document I posted (obtained via output { file { ... }})

**Comment by jsvd** [comment]
@darkmoon03 unfortunately the events you posted don't seem to help much as they contain only the actual event data. When this data is sent to elasticsearch as bulk request there's a lot of metadata that drives the indexing of the document, and it's there that seems to lie the issue.

Can you confirm that `retry_on_conflict` is being set to a high value? if you start the pipeline with `--log.level=debug` the plugin options are printed to the logs.

The next thing we can try is to patch your es output in your logstash installation to print the bulk request before it is sent.

**Comment by darkmoon03** [comment]
Yes, retry_on_conflict is set to a high value.  I've had it as high as 100.  It only makes things worse.  (all the retries fail.)

The problem is probably in the meta data - If I send the same event by API, it works as expected.

**Comment by darkmoon03** [comment]
How would we patch the es output to print the bulk request?

*IS* it a bulk request?  I can get this to happen on a single document update.

**Comment by darkmoon03** [comment]
bump

**Comment by vhdm** [comment]
> No solution to this? I'm having the same trouble.

hi, can you find any solution for this problem?
